refactor!: move source link creation, track state creation, measurement selection into a single unit outside of the CKF

goetzgaycken opened the pull request:
Merge retrieval of source link ranges, measurement selection and
track state creation into one unit which the CKF interacts with
via a single delegate. The delegates for measurement selection
track state creation and calibration are removed from the CKF
options/extensions.
    
The original building blocks (SourceLink accessor, measurement selector,
and track state creator, or the combined measurement selector and track
state creator) can still be used, however require to setup an additional
helper object which combines these independent components. The
algorithmic code is unchanged.

As a prerequisite the track states are converted to bound states also on
empty sensitive surfaces.

This increases the computational effort slightly for empty, sensitive
surfaces, since the computation of bound states is slightly more
demanding than the computation of curvilinear states, but for complex
events like HL-LHC events the degradation was not measurable.

--- END COMMIT MESSAGE ---

Any further description goes here, @-mentions are ok here!


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced `TrackStateCreator` for enhanced track state management.
  - Added `CombinatorialKalmanFilterExtensions` for improved filter behavior customization.

- **Improvements**
  - Simplified method signatures in `CombinatorialKalmanFilterOptions`, reducing complexity.
  - Enhanced modularity in `TrackFindingAlgorithm` and related components.
  - Improved clarity and maintainability in the test framework.
  - Streamlined connections for measurement selection and calibration within the track finding process.

- **Bug Fixes**
  - Streamlined error handling and measurement selection processes.

- **Documentation**
  - Updated documentation generator to include new structures for better tracking.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

goetzgaycken commented:
- rebase to main
- resolved merge conflicts with: fix: protect TrackSelector against eta=nan (#3785)

goetzgaycken commented:
changes d3a565e29b216f8ed7c6f69c89edb09c35a84a68 to b44b8fe89abae78002c327ff462b54ed7828be29:
- fix doxygen template parameter tag

goetzgaycken commented:
Why should there be brackets arround slAccessorDelegate [TrackFindingAlgorithm.cpp:365](https://github.com/acts-project/acts/blob/b44b8fe89abae78002c327ff462b54ed7828be29/Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp#L364-367) (macos build complains):
```
364 | TrackStateCandidateCreatorType trackStateCreator{
365 |    slAccessorDelegate,
366 |    Acts::makeDelegate<&MeasurementCalibratorAdapter::calibrate>(calibrator),
367 |    Acts::makeDelegate<&MeasurementSelector::select>(measSel)};
```

paulgessinger commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+      // CAREFUL! This trackstate has a previous index that is not in this
+      // MultiTrajectory Visiting brackwards from this track state will
+      // fail!
+      auto ts = bufferTrajectory.makeTrackState(mask, prevTip);
```
The only remaining reason for this `bufferTrajectory` is to allow the use of the previous measurement selector after calibration, which requires track states, right?

goetzgaycken replied:
- Removed buffer trajectory argument (in the only use case, the buffer trajectory and the main trajectory had been the same).
- Moved the buffer for newly created candidate track state proxies to the TrackStateCreator, where it is now a mutable member.

goetzgaycken replied:
moved temporary buffer for track state candidates back to the CKF results, which is now passed again by argument to the TrackStateCreator

paulgessinger commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+  };
+};
+
+/// @brief Base class for the ComposableTrackStateCreator which simply delegates its tasks
```
This is now derived, not base anymore, isn't it?

goetzgaycken replied:
exactly.  Thanks!

paulgessinger commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+
+  SourceLinkAccessor sourceLinkAccessor;
+
+  const track_state_creator_derived_t& derived() const {
```
This could be private maybe.

goetzgaycken replied:
Not sure whether that is something that needs to be hidden from the public, but it is presumably also not really necessary  to keep it public:

andiwand replied:
IMO it is always a good idea to keep the public surface as small as possible. I don't think this is useful to the user?

goetzgaycken replied:
done

paulgessinger commented on Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:
```diff
@@ -370,8 +395,11 @@ BOOST_AUTO_TEST_CASE(ZeroFieldBackward) {
-  options.sourceLinkAccessor.connect<&Fixture::TestSourceLinkAccessor::range>(
-      &slAccessor);
+
+  auto trackStateCreator = makeTrackStateCreator(slAccessor, f.measSel);
```
Is this used anywhere?

goetzgaycken replied:
Yes, that is the delegate for extendOrBranchTrajectory. Maybe I should rename it ?

goetzgaycken replied:
name now agrees with the delegate name.

paulgessinger commented:
Couple more thoughts on this:

I think going with a single level of composition would be clearer. Currently there's the base composable class that connects to the previous `SourceLinkAccessor`, and then a second layer which accepts the other original delegates. 

A single *composable* struct that by itself has the three (?) original delegates including the `SourceLinkAccessor` would be sufficient if we drop the compatibility with the intermediate *track state creator* style delegate. 

What do we need this second layer for after this change? I think having one layer of adapter to the fully separate makes sense, but the surface area of the *track state creator* seems small enough to me to remove that altogether.

goetzgaycken commented:
It  is only needed for the transition period. In my opinion it would be better to first introduce the new api and then merge on the athena side the track state creator and source link accessor. The amount of code duplication when introducing different helper objects for the two cases is small, though the simplification is also not significant. Therefore I  would rather not duplicate the code, and rather remove  the unnecessary code and layers later when they are not needed anymore. But I do not have a strong opinion.

goetzgaycken commented on Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:
```diff
@@ -357,15 +352,33 @@ ProcessCode TrackFindingAlgorithm::execute(const AlgorithmContext& ctx) const {
+  using TrackStateCandidateCreatorType =
+      Acts::TrackStateCandidateCreatorImpl<IndexSourceLinkAccessor::Iterator,
+                                           TrackContainer>;
+  TrackStateCandidateCreatorType trackStateCreator{
+      slAccessorDelegate,
+      Acts::makeDelegate<&MeasurementCalibratorAdapter::calibrate>(calibrator),
+      Acts::makeDelegate<&MeasurementSelector::select>(measSel)};
```
You create a not properly initialized object, and the compiler would not tell You if You missed something.

paulgessinger replied:
Since these are supposed to be runtime configurable, at least optionally, I don't see this as an issue. What we do everywhere else is either introduce default functions (which you do) than can error on invocation, or when pushing the object to a consumer it can check if all of the delegates are connected.

goetzgaycken replied:
I don't have a strong opinion. But if it is preferred I can remove these helper methods and the starting point would be an uninitialized object and it is up to the user to remember what needs to be connected.

Though I don't see it as an improvement to move something that could be caught during compile time to the runtime. But of course You are right that delegates could be fully configurable thus it often is a runtime problem anyways.

It also seems to simplify client code. There are a few cases like 
```
Delgate delegate;
delegate.connect<&func>(obj):
Options option(delegate);
```
which just becomes 
```
Options option(makeDelegate<&func>(obj));
```
Of course the first version could just be 
```
Options option;
option.delegate.connect<&func>(obj):
```
But then the delegate has to get a default value also in cases in which there is no meaningful default value, and in which the default value is only short lived.

paulgessinger replied:
I would prefer this, yes. There's also the added benefit that explicitly writing the delegate being connected avoids any ordering confusion (although this is a compile error due to mismatching types).

goetzgaycken replied:
I don't understand the argument about ordering confusion, either way You cannot make a "connected" delegate to a non existing object or function by accident, or at least I do not see how one snippet is more likely for errors than the other, except for uninitialized member delegates which are caught at runtime in one case and at compile time in the other. 

But I see You prefer for some reason unused default values and or more boiler plate code.

andiwand commented:
The question for me would be, do we need a transition period? The change is breaking either way no? Touching this twice means we have to go through two breaking cycles.

goetzgaycken commented:
> The question for me would be, do we need a transition period? The change is breaking either way no? Touching this twice means we have to go through two breaking cycles.

It is certainly possible to merge at the same time on the athena side the source link accessor and track state creator. I don't have any strong argument to motivate why it is better to break this into smaller steps.

goetzgaycken commented on Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:
```diff
@@ -357,15 +352,33 @@ ProcessCode TrackFindingAlgorithm::execute(const AlgorithmContext& ctx) const {
+                                   secondPropOptions);
   secondOptions.targetSurface = m_cfg.reverseSearch ? nullptr : pSurface.get();
   secondOptions.skipPrePropagationUpdate = true;
+  secondOptions.extendOrBranchTrajectory =
```
At the moment the extensions struct would only contain a delegate for the updater and branchStopper. Both are optional, aren't they? And in theory could be identical for the entire job. In practice e.g. the branchStopper might not be stateless (e.g. information whether it is the in-side-out or out-side-in pass). The extendOrBranchTrajectory  has to know at least something about the current event, so cannot be stateless (not considering some thread local global variables ... )

It is not fully clear to me what the role of the extensions should be or become. At the moment the extensions seem to be a bit of an arbitrary subset of the options. But if it makes sense to collect all delegates in one struct and everything which is not a delegate stays in the parent struct then certainly the extendOrBranchTrajectory should be moved to the extensions.

There also was the question about the naming. Is this a good name ? Or is it misleading ? It does add either one or more new track states to the current branch (which still need some post processing), so it should do what the name suggests.

goetzgaycken replied:
renamed to trackStateCreator

goetzgaycken commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:
```diff
@@ -1098,22 +818,13 @@ class CombinatorialKalmanFilter {
     /// @note will be set to a default trackStateCandidateCreator or the one
     //        provided via the extension
-    TrackStateCandidateCreator trackStateCandidateCreator;
+    TrajectoryExtender extendOrBranchTrajectory;
```
At the moment none of the members in options or extensions has the "m_" prefix. In that sense adding an "m_" prefix would make it inconsistent.

goetzgaycken replied:
did not add "m_" prefix:
- it is a struct
- contains only type definitions and delegates; without "m_" prefix invoking delegates look like function calls.

goetzgaycken commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+      std::vector<TrackStateProxy>& trackStateCandidates,
+      TrackStateContainerBackend& trajectory, const Logger& logger) const {
+    TrackStatesResult tsRes = TrackStatesResult::success({});
+    using SourceLinkRange = decltype(sourceLinkAccessor(surface));
+    std::optional<SourceLinkRange> slRange = sourceLinkAccessor(surface);
```
That is copy&pasted from the original code in the CKF the optional lost its purpose. So should be removed. the sourceLinkAccessor returns a range not an optional.

goetzgaycken replied:
I did remove the optional

goetzgaycken commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+//    data which is to be filled by the next stage e.g. the
+//    CombinatorialKalmanFilter.
+template <typename source_link_iterator_t, typename track_container_t>
+struct TrackStateCandidateCreatorImpl
```
Maybe the class should be renamed to fit the delegate name. In the original implementation in which the hierarchy was inverted this was the base class, and then this name made more sense.

goetzgaycken replied:
renamed to TrackStateCreator

andiwand commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+
+}  // namespace CkfTypes
+
+namespace detail {
+// helper method to deduce the return type and arguments of the member function
+template <auto memberfunc_ptr, typename return_t, typename instance_t,
+          typename... Args>
+inline auto makeDelegate(
+    [[maybe_unused]] return_t (instance_t::*memberfunc)(Args...) const,
+    const instance_t& instance) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<memberfunc_ptr>(&instance);
+  return delegate;
+}
+// helper method to deduce the return type and arguments of the function
+template <auto func_ptr, typename return_t, typename... Args>
+inline auto makeDelegate([[maybe_unused]] return_t (*func)(Args...)) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<func_ptr>();
+  return delegate;
+}
+}  // namespace detail
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam memberfunc_ptr a pointer to a member function of the class instance_t
+/// @param instance an instance which outlives this delegate.
+template <auto memberfunc_ptr, typename instance_t>
+inline auto makeDelegate(const instance_t& instance) {
+  return detail::makeDelegate<memberfunc_ptr>(memberfunc_ptr, instance);
+}
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam func_ptr a pointer to a function
+template <auto func_ptr>
+inline auto makeDelegate() {
+  return detail::makeDelegate<func_ptr>(func_ptr);
+}
```
This seems a bit generic for a header called `ComposableTrackStateCreator.hpp`. Either this should be a utility placed in the delegate header or dropped in favor of doing this by hand.

I think the deduction should also work without a function wrapper?

goetzgaycken replied:
I removed these helper functions, although I think that they would have been useful

goetzgaycken commented:
changes from  2cc2ecc to dbfba96 :
- attempt to solve unresolved doxygen reference.

goetzgaycken commented:
changes from 2fa962c to 9493657:
- rebase to main of 2024-11-28_T1014

goetzgaycken commented on Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp:
```diff
@@ -0,0 +1,415 @@
+/// stateCandidateCreator
+/// @note if the number of states exceeds this number dynamic memory allocation will occur.
+///       the number is chosen to yield a container size of 64 bytes.
+static constexpr std::size_t s_maxBranchesPerSurface = 10;
```
Resolved, the definition is now in the CombinatorialKalmanFilterExtensions.hpp which defines the delegate which needs this definition.

goetzgaycken commented:
Open discussion points:
- should the new extension member be prefixed with "m_" although non of the others are prefixed in this way ? Extend this PR and add the prefix for all members ?
- is the name "extendOrBranchTrajectory" a good name for the delegate ? Or should it be called trackStateCreator i.e the delegate this is kind of replacing (together with the sourceLinkAccessor)
- Remove TrackStateCreatorDelegate which provides a helper which allows to use  the previous track state create interface currently used by athena ? This would require to directly implement the new interface on the athena side.
- Rename TrackStateCandidateCreatorImpl ? To what ?
- Remove the buffer trajectory etc. from the new interface ? This would require that the implementation of the ComposableTrackStateCreator  (TrackStateCandidateCreatorImpl) would have to provide this i.e. would gain this as a mutable member.

goetzgaycken commented:
changes from 9493657 to abedde4:
- added missing include file, which only was included indirectly before.

andiwand commented:
> Open discussion points:

my 2 cents

> * should the new extension member be prefixed with "m_" although non of the others are prefixed in this way ? Extend this PR and add the prefix for all members ?

The lose guideline here is that for `struct`s we don't use `m_` but only for classes.

> * is the name "extendOrBranchTrajectory" a good name for the delegate ? Or should it be called trackStateCreator i.e the delegate this is kind of replacing (together with the sourceLinkAccessor)

I would be in favor of keeping `trackStateCreator`

> * Remove TrackStateCreatorDelegate which provides a helper which allows to use  the previous track state create interface currently used by athena ? This would require to directly implement the new interface on the athena side. The idea is that e.g. the minimal changes to support the new interface would be simpler: ([cern-gitlab:f8ac37ab3f](https://gitlab.cern.ch/goetz/athena/-/commit/f8ac37ab3f73ac693b7fbdf6b31ad350af4091c1), to directly support the new interface more changes are needed:  [cern-gitlab:171095fc26](https://gitlab.cern.ch/goetz/athena/-/commit/171095fc26884755d67131979580d300cfa68fd5)

To me it seems easier to make one sharp cut and break the API directly without a transition phase. I think Atlas is the only experiment using this delegate so it is really up to you IMO. What is easier for you?

> * Rename TrackStateCandidateCreatorImpl ? To what ?

I wondered if we remove the compatibility layer for the existing `TrackStateCreator` if we could/should collapse `ComposableTrackStateCreator` and `TrackStateCandidateCreatorImpl` into one class?

> * Remove the buffer trajectory etc. from the new interface ? This would require that the implementation of the ComposableTrackStateCreator  (TrackStateCandidateCreatorImpl) would have to provide this i.e. would gain this as a mutable member.

I think this would be good to remove. In case we drop the existing `TrackStateCreator` this would come for free?

goetzgaycken commented:
> > * should the new extension member be prefixed with "m_" although non of the others are prefixed in this way ? Extend this PR and add the prefix for all members ?
> 
> The lose guideline here is that for `struct`s we don't use `m_` but only for classes.
> 

advantage "m_"  prefix: 
- easy to see what is a member variable and what not without needing an IDE
disadvantage "m_" prefix:
- the CKF extensions is a struct so would be against the convention
- delegate calls don't look like function calls but have the "m_" prefix
I presumably would not the prefix because there wasn't one.

> > * is the name "extendOrBranchTrajectory" a good name for the delegate ? Or should it be called trackStateCreator i.e the delegate this is kind of replacing (together with the sourceLinkAccessor)
> 
> I would be in favor of keeping `trackStateCreator`

ok.

> > * Remove TrackStateCreatorDelegate which provides a helper which allows to use  the previous track state create interface currently used by athena ? This would require to directly implement the new interface on the athena side. The idea is that e.g. the minimal changes to support the new interface would be simpler: ([cern-gitlab:f8ac37ab3f](https://gitlab.cern.ch/goetz/athena/-/commit/f8ac37ab3f73ac693b7fbdf6b31ad350af4091c1), to directly support the new interface more changes are needed:  [cern-gitlab:171095fc26](https://gitlab.cern.ch/goetz/athena/-/commit/171095fc26884755d67131979580d300cfa68fd5)
> 
> To me it seems easier to make one sharp cut and break the API directly without a transition phase. I think Atlas is the only experiment using this delegate so it is really up to you IMO. What is easier for you?

Certainly reduces the amount of new code.

> > * Rename TrackStateCandidateCreatorImpl ? To what ?
> 
> I wondered if we remove the compatibility layer for the existing `TrackStateCreator` if we could/should collapse `ComposableTrackStateCreator` and `TrackStateCandidateCreatorImpl` into one class?

Indeed, if there is only a single implementation of ComposableTrackStateCreator there would not be a need to split off a derived class.

> > * Remove the buffer trajectory etc. from the new interface ? This would require that the implementation of the ComposableTrackStateCreator  (TrackStateCandidateCreatorImpl) would have to provide this i.e. would gain this as a mutable member.
> 
> I think this would be good to remove. In case we drop the existing `TrackStateCreator` this would come for free?

The Acts measurement selector needs the buffer trajectory, so it would be moved from the CKF result  and become a member of the track state creator helper class (TrackStateCandidateCreatorImpl). Interfaces become simpler. The TrackStateCandidateCreatorImpl would get a mutable state, but already has the source link accessor, so has an event state anyways.

goetzgaycken commented:
Changes from abedde4 to df55dfb:
- removed support for the previous track state creator interface
- replaced composable track state creator by its single implementation, and renamed file
- renamed delegate etc. to trackStateCreate
- removed buffer trajectory, and track state candidate  argument and, moved vector for candidate track state proxies to the helper class (mutable member)

Updated changes an the athena side can be found here [cern-gitlab:main_support_for_new_extendOrBranchTrajectory_delegate)](https://gitlab.cern.ch/goetz/athena/-/commits/main_support_for_new_extendOrBranchTrajectory_delegate)

goetzgaycken commented:
changes  df55dfb to 9261956:
- fixed spelling in comment

goetzgaycken commented:
Changes from 9261956 to cae4ee3:
- added missing include
- forgot to propagate name change to parts of the documentation

goetzgaycken commented:
Still an open question:

is the mutable member in the TrackStateCreator for the track state candidate proxies a good idea ?

 In case track finding is executed in parallel each thread (or even coroutine) would need its own version of the TrackStateCreator. It may sound worse than it is, since the trajectory cannot be extended concurrently, and the TrackStateCreator contains also the source link accessor which must have some kind of event state. 
 
Though, the latter are not unsolvable, the trajectory is passed as argument to the track finder, and the source link accessor could get the event state from thread local storage. 

But, on the other hand is good to have an argument in the trackStateCreator delegate and a buffer in CKF Result which has no obvious purpose except for the one implementation of the trackStateCreator which is currently in Acts ?  

Maybe it is better to delay the removal until there is an implementation in Acts which does not need this buffer (?)

andiwand commented:
I think you are right - this kind of hidden thread local state is what we usually try to avoid in ACTS so I agree with you that it might be best for now to carry it through. I was hoping to get rid of it already but forgot about this side effect.

goetzgaycken commented:
changes  cae4ee3 to 56d0051:
- moved the buffer for candidate track states back to the CKF results and pass it as argument to trackStateCreator.

corresponding update on the athena side is here:  [main_support_for_new_extendOrBranchTrajectory_delegate](cern-gitlab:](https://gitlab.cern.ch/goetz/athena/-/commits/main_support_for_new_extendOrBranchTrajectory_delegate)

goetzgaycken commented:
I don't have any further ideas. Undrafting.

goetzgaycken commented:
A bit misleading summary by coder abitai, 
- it correctly understood that it was about restructuring of parts of the CKF,
- it correctly understood that  method signatures are simplified,
- it correctly "understood" that the molecularity improved i.e. better decoupling of CKF and source links, track state creation.
- But what is listed under "bug fixes" and "new features" is incorrect.

goetzgaycken commented:
changes 56d0051 to dca3186:
- adjust ActsExamplesTrackFindingExaTrkX for new track state creator interface

andiwand commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:
```diff
@@ -39,97 +40,12 @@
 /// @tparam source_link_iterator_t Type of the source link iterator
 /// @tparam track_container_t Type of the track container
-template <typename source_link_iterator_t, typename track_container_t>
+template <typename track_container_t>
```
down to one template argument 🎉

andiwand commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:
```diff
@@ -309,190 +186,13 @@ class CombinatorialKalmanFilter {
   /// The CombinatorialKalmanFilter Actor does not rely on the measurements to
   /// be sorted along the track.
-  template <typename source_link_accessor_t, typename parameters_t>
+  template <typename parameters_t>
```
I wonder if we could drop that? But that is something for another PR

goetzgaycken replied:
I actually wonder whether this can be different from BoundTrackParameters

andiwand replied:
Right I think this can only be `BoundTrackParameters` since it should not be coupled to the start parameters. Even for the start parameters I would argue that only `BoundTrackParameters` make sense for now

goetzgaycken replied:
Something for a different PR

andiwand commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:
```diff
@@ -1161,21 +829,17 @@ class CombinatorialKalmanFilter {
   /// @return a container of track finding result for all the initial track
   /// parameters
-  template <typename source_link_iterator_t, typename start_parameters_t,
+  template <typename start_parameters_t,
```
These two template args also look like relics. We can clean that up later

andiwand commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilterExtensions.hpp:
```diff
@@ -0,0 +1,99 @@
+
+namespace Acts {
+
+/// expected max number of track states that are expected to be added by
+/// stateCandidateCreator
+/// @note if the number of states exceeds this number dynamic memory allocation will occur.
+///       the number is chosen to yield a container size of 64 bytes.
+static constexpr std::size_t s_maxBranchesPerSurface = 10;
```
Since we are breaking the API, do you think it makes sense to move this into a CKF related namespace?

goetzgaycken replied:
We can move this to the CfkTypes namespace or even move this and the CfkTypes to the extensions.

goetzgaycken replied:
Just experimenting with this. The  s_maxBranchesPerSurface  can certainly be moved into a namespace. I said one can move everything into the CKF extensions. The disadvantage would be that the full type name becomes a template of the track container and is very long. Thus, aliases would be used where ever these types appear. These aliases makes it often tedious to search and/or understand the actual definition of a type.  Anyway I would not have a strong opinion.

andiwand replied:
Yes I agree we should not put it behind one of the templated classes. Potentially `CkfTypes` makes sense but I don't have a strong preference. I just thought `s_maxBranchesPerSurface` is a bit generic for the `Acts` namespace.

goetzgaycken replied:
moved s_maxBranchesPerSurface into CkfTypes

andiwand commented on Core/include/Acts/TrackFinding/TrackStateCreator.hpp:
```diff
@@ -0,0 +1,294 @@
+/// measurements, are created in the given trajectory. The resulting container
+/// may become big. Thus, it is advisable to copy selected tracks and their
+/// track states to a separate container after each track finding step.
+
```
```suggestion
///
```
nitpick

andiwand reviewed (dismissed):
This is great!

I left a few minor comments.

Since this is breaking we can only merge this in ~2 weeks. But I don't expect it to conflict with other CKF developments.

goetzgaycken commented:
changes 7d5548c to 3639a5a:
- ... and the 3rd attempt to fix the delegate connections for the ExaTrk proto track algorithm

goetzgaycken commented:
changes 3639a5a to 335974e:
- fix setup of track finder options of ExaTrk proto track finding

goetzgaycken commented on Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:
```diff
@@ -289,14 +284,40 @@ struct Fixture {
   }
 };
 
+// set up composable track state creator from these components:
+//  - source link accessor,
+//  - measurement selector
+//  - track  state candidate creator
+template <typename source_link_accessor_t>
+inline auto makeTrackStateCreator(const source_link_accessor_t& slAccessor,
+                                  const Acts::MeasurementSelector& measSel) {
+  using TrackStateCreatorType =
+      Acts::TrackStateCreator<typename source_link_accessor_t::Iterator,
+                              TrackContainer>;
+  TrackStateCreatorType trackStateCreator;
+  trackStateCreator.sourceLinkAccessor
+      .template connect<&source_link_accessor_t::range>(&slAccessor);
+  trackStateCreator.calibrator.template connect<
+      &testSourceLinkCalibrator<TrackStateContainerBackend>>();
+  trackStateCreator.measurementSelector.template connect<
+      &Acts::MeasurementSelector::select<TrackStateContainerBackend>>(&measSel);
+  return trackStateCreator;
+}
```
Where does master coder abitai get this from ?

goetzgaycken replied:
Dear master coder abitai, unfortunately connect returns void. So, Your error handing is not possible.

goetzgaycken commented on Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:
```diff
@@ -318,27 +319,34 @@ ProcessCode TrackFindingAlgorithm::execute(const AlgorithmContext& ctx) const {
   using Extensions = Acts::CombinatorialKalmanFilterExtensions<TrackContainer>;
 
   BranchStopper branchStopper(m_cfg);
+  MeasurementSelector measSel{
+      Acts::MeasurementSelector(m_cfg.measurementSelectorCfg)};
+
+  IndexSourceLinkAccessor slAccessor;
+  slAccessor.container = &measurements.orderedIndices();
+
```
Dear mater coder abitai, I think an empty measurement range is acceptable. Anyway has nothing to do with this PR.

goetzgaycken commented on Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:
```diff
@@ -318,27 +319,34 @@
-  Acts::SourceLinkAccessorDelegate<IndexSourceLinkAccessor::Iterator>
-      slAccessorDelegate;
-  slAccessorDelegate.connect<&IndexSourceLinkAccessor::range>(&slAccessor);
+  extensions.createTrackStates
+      .template connect<&TrackStateCreatorType ::createTrackStates>(
+          &trackStateCreator);
```
Dear master coder abitai, I think connect returns void .

goetzgaycken commented on Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:
```diff
@@ -318,27 +319,34 @@
+  IndexSourceLinkAccessor slAccessor;
+  slAccessor.container = &measurements.orderedIndices();
+
+  using TrackStateCreatorType =
+      Acts::TrackStateCreator<IndexSourceLinkAccessor::Iterator,
+                              TrackContainer>;
+  TrackStateCreatorType trackStateCreator;
+  trackStateCreator.sourceLinkAccessor
+      .template connect<&IndexSourceLinkAccessor::range>(&slAccessor);
+  trackStateCreator.calibrator
+      .template connect<&MeasurementCalibratorAdapter::calibrate>(&calibrator);
+  trackStateCreator.measurementSelector
+      .template connect<&MeasurementSelector::select>(&measSel);
```
still returns void

goetzgaycken commented:
Changes from  335974e to 175a588:
- rebase to main of 2024-12-04_T1314
- propagated Juan Miguel Carceller's spelling fix to the TrackStateCreator 
- moved definition of s_maxBranchesPerSurface into namespace CkfTypes

goetzgaycken commented:
Changes from 175a588 to 5dd2547:
- added the "lost the /// instead of an empty line" fix

goetzgaycken commented on Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:
```diff
@@ -768,17 +454,17 @@ class CombinatorialKalmanFilter {
-        tsRes = trackStateCandidateCreator(
+      if (isSensitive) {
+        // extend trajectory with measurements associated to the current surface
+        // which may create extra trajectory branches if more than one
+        // measurement is selected.
+        tsRes = m_extensions.createTrackStates(
             state.geoContext, *calibrationContextPtr, *surface, boundState,
-            slBegin, slEnd, prevTip, *result.trackStates,
-            result.trackStateCandidates, *result.trackStates, logger());
+            prevTip, result.trackStateCandidates, *result.trackStates,
+            logger());
```
Dear master coder abitai,
The  proposed debug message is useless since the cache is expected to be empty before calling createTrackStates. But the proposal for the error message makes indeed sense.

goetzgaycken replied:
Applied improvement to error message.

goetzgaycken commented:
Changes from 5dd2547 to 65948cf:
- improved error message as suggested by coderabbitai

Changes from [65948cf](https://github.com/acts-project/acts/commit/65948cf3cbd7c960fe4debdb6d851da168475f58) to [dbd9d5e](https://github.com/acts-project/acts/commit/dbd9d5ea71593a0e0981affa2d8712c30bc27c97):
- rebase to main of 2024-12-10_T2146

goetzgaycken commented:
changes:
[dbd9d5e](https://github.com/acts-project/acts/commit/dbd9d5ea71593a0e0981affa2d8712c30bc27c97) to [b8e7d2b](https://github.com/acts-project/acts/commit/b8e7d2ba8937ce533c989f1144c0e709b8ae7aeb):
- white space fixes

goetzgaycken commented:
The corresponding necessary changes for athena can be found in [gitlab-cern:main_support_for_new_extendOrBranchTrajectory_delegate](https://gitlab.cern.ch/goetz/athena/-/commits/main_support_for_new_extendOrBranchTrajectory_delegate)

paulgessinger commented:
Let's make sure to get this in for v39.

goetzgaycken commented on Core/include/Acts/TrackFinding/TrackStateCreator.hpp:
```diff
@@ -0,0 +1,294 @@
+
+      // prepare the track state
+      PM mask = PM::Predicted | PM::Jacobian | PM::Calibrated;
+      if (it != slBegin) {
+        // not the first TrackState, only need uncalibrated and calibrated
+        mask = PM::Calibrated;
+      }
```
it is intended to overwrite the mask.
